libdrgn/python/drgnpy.h

```diff
@@ -97,6 +97,14 @@ typedef struct {
 	struct pyobjectp_set objects;
 } Program;
 
+typedef struct _GenericIterator {
```

Comment:
If it's not too difficult to change, after seeing the code I think I'd actually prefer what you mentioned about unique types for each iterator and generated the boilerplate with macros.

Reply:
I've started implementing this, but I _really_ don't think you're going to like it. The macro (ab)use necessary to do this without a ton of boilerplate is IMO _far_ more unwieldy than the approach with function pointers and `void*`.

As far as when I initially asked, I was actually referring to making the underlying C iterators concrete (i.e. `radix_tree_iter`, `idr_iter`, `pid_iter`, and `task_iter`) or having them be implemented in terms of function pointers and `void*`.  My logic was that on the Python side things are already so dynamic that the additional burden of making the corresponding Python iterators distinct types wasn't worth it (especially since they'd ultimately be implementation details that wouldn't be directly exposed to the user, much like `ThreadIterator`). I can put up my commit with the concrete macro-defined iterators if you want, but again I really don't think you're going to prefer this approach once you see it. Happy to do it though, either way just let me know.

Reply:
Alright, let's stick with the function pointer way.